onnxruntime/core/optimizer/qdq_transformer/qdq_op_transformer.cc

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "qdq_op_transformer.h"
+
+#include <string>
+#include <vector>
+
+#include "core/graph/graph.h"
+#include "core/graph/onnx_protobuf.h"
+
+namespace onnxruntime {
+
+void QDQOperatorTransformer::FillQDQOptionalZeroPoint(const std::vector<const Node*>& qdq_nodes) {
+  for (const Node* p_node_const : qdq_nodes) {
+    Node& node = *graph_.GetNode(p_node_const->Index());
+    std::vector<NodeArg*>& input_defs = node.MutableInputDefs();
+    if (input_defs.size() == 3) continue;
+
+    bool is_default_zp_signed = false;
+    if (node.OpType() == DQOPTypeName) {
+      auto input_type = input_defs[0]->TypeAsProto()->tensor_type().elem_type();
+      is_default_zp_signed = ONNX_NAMESPACE::TensorProto_DataType_INT8 == input_type;
+    }
+
+    const ONNX_NAMESPACE::TensorProto& zp_tensor_proto = is_default_zp_signed ? optional_zero_point_int8_ : optional_zero_point_uint8_;
+
+    const ONNX_NAMESPACE::TensorProto* dummy_zp_tensor_proto;
+    if (!graph_.GetInitializedTensor(zp_tensor_proto.name(), dummy_zp_tensor_proto)) {
+      graph_.AddInitializedTensor(zp_tensor_proto);
+    }
+
+    input_defs.push_back(&graph_.GetOrCreateNodeArg(zp_tensor_proto.name(), nullptr));
```

Comment:
Could we just add these at the start and not bother checking here? Graph::Resolve will throw away anything that isn't used later on.

Reply:
The datatype of zero_point of DequantizeLinear is determined by its 1st input. The check is mainly for that.

Reply:
Sorry - I could have been clearer. I meant add both the int8 and uint8 zero points at the start as Graph::Resolve with throw anything unused away. Optional to change.

---
In reply to: [600846628](https://github.com/microsoft/onnxruntime/pull/7033#discussion_r600846628) [](ancestors = 600846628)